core/restorer.go

```diff
@@ -79,3 +79,69 @@ func (r *Restorer) CheckSecondaryControllerNodes() map[string]error {
 	}
 	return reachable
 }
+
+// StopAgents stops controller agents, jujud-machine-*.
+// If stopSecondaries is true, these agents on other controller nodes will be stopped
+// as well.
+// The agents on the primary node are always stopped last.
+func (r *Restorer) StopAgents(stopSecondaries bool) map[string]error {
+	return r.manageAgents(stopSecondaries, func(n ControllerNode) error { return n.StopAgent() })
+}
+
+// StartAgents starts controller agents, jujud-machine-*.
+// If stopSecondaries is true, these agents on other controller nodes will be started
+// as well.
+// The agents on the primary node are always started first.
+func (r *Restorer) StartAgents(startSecondaries bool) map[string]error {
+	// Check replicaset is healthy before restarting agents.
+	r.replicaSetStabilised()
+	return r.manageAgents(startSecondaries, func(n ControllerNode) error { return n.StartAgent() })
+}
+
+// TODO Need to figure out how to wait for stability....
+// Is this enough?..
+func (r *Restorer) replicaSetStabilised() {
+	// We want to refresh replicaset as we go...
+	attempt := 0
+	// keep a copy of replicaset, just in case
+	pre := r.replicaSet
+	for {
+		attempt++
+		logger.Debugf("checking db health, attempt %d", attempt)
+		replicaSet, err := r.db.ReplicaSet()
+		if err != nil {
+			logger.Errorf("getting database replica set: %v", err)
+			continue
+		}
+		r.replicaSet = replicaSet
+		err = r.CheckDatabaseState()
+		if err == nil {
+			logger.Debugf("replicaset is healthy again")
+			break
+		} else {
+			logger.Debugf(" replicaset is sick: %v", err)
+		}
+		if attempt == 20 {
+			r.replicaSet = pre
+			logger.Debugf("Could not finish waiting for healthy replicaset")
+			break
+		}
+	}
+}
+
+func (r *Restorer) manageAgents(all bool, operation func(n ControllerNode) error) map[string]error {
+	var primary ControllerNode
+	result := map[string]error{}
+	for _, member := range r.replicaSet.Members {
+		memberMachine := r.convertToControllerNode(member)
+		if member.Self {
+			primary = memberMachine
+			continue
+		}
+		if all {
+			result[memberMachine.IP()] = operation(memberMachine)
+		}
+	}
+	result[primary.IP()] = operation(primary)
```

Comment:
Hmm.. you got me thinking... What we actually want is to stop primary last (to avoid last minute re-election) but start it first (more chances it will be re-elected primary again since it'll likely be the first to ask to be  a primary). 

Changed the code slightly for it :)